memoryview: slice assignment RFC

dxxb opened the pull request:
Hi,

I'd like to copy memoryview slices without resorting to copying byte by byte and found that things like memview2[1:3] = memview1[0:2] do not currently work so I took a stab at adding support for it. Before applying the PR, the augmented memoryview test below results in the following output:

```
# ../unix/micropython basics/memoryview1.py
4
49 50 52
[49, 50, 51, 52]
TypeError
bytearray(b'\x01234')
[1, 50, 51, 52]
[50, 51, 52]
[50, 51]
bytearray(b'\x00\x00')
[1, 2, 3, 4]
[2, 3]
array('i', [1, 2, 6, 4])
array/bytes required on right side
array/bytes required on right side
lhs and rhs should be compatible
lhs and rhs should be compatible
array/bytes required on right side
```

Applying the PR I get:

```
# ../unix/micropython basics/memoryview1.py
4
49 50 52
[49, 50, 51, 52]
TypeError
bytearray(b'\x01234')
[1, 50, 51, 52]
[50, 51, 52]
[50, 51]
bytearray(b'\x00\x00')
[1, 2, 3, 4]
[2, 3]
array('i', [1, 2, 6, 4])
bytearray(b'5128')
bytearray(b'5128')
bytearray(b'1212')
ValueError
ValueError
```

For reference Python3.4 output matches:

```
# python3.4 basics/memoryview1.py                                                                                                                             
4
49 50 52
[49, 50, 51, 52]
TypeError
bytearray(b'\x01234')
[1, 50, 51, 52]
[50, 51, 52]
[50, 51]
bytearray(b'\x00\x00')
[1, 2, 3, 4]
[2, 3]
array('i', [1, 2, 6, 4])
bytearray(b'5128')
bytearray(b'5128')
bytearray(b'1212')
ValueError
ValueError
```

Comments?

dxxb committed:
py: memoryview slice assignment

dxxb committed:
py: fix indent error

dxxb committed:
qemu-arm: enable array slice assign in port

dxxb committed:
py: fixup additions to memoryview tests

pfalcon commented:
Thanks for the patch. Please show how it'll work on:

~~~~
a = array.array('i', [0]*5)
m = memoryview(a)
~~~~

pfalcon commented on py/objarray.c:
```diff
@@ -405,12 +415,17 @@ STATIC mp_obj_t array_subscr(mp_obj_t self_in, mp_obj_t index_in, mp_obj_t value
-                    // TODO: alloc policy after shrinking
+                    if (o->base.type == &mp_type_memoryview) {
+                        // destination is memoryview use memcpy() instead of mp_seq_replace_slice_no_grow() below
+                        memcpy((uint8_t*)o->items + o->free + slice.start, src_items, src_len);
```
So, this comment states the obvious. Instead, it should tell why everything uses mp_seq_replace_slice_no_grow(), and memoryview suddenly doesn't.

dxxb replied:
mp_seq_replace_slice_no_grow() calls memcpy() and memmove(), I wanted to avoid calling memmove() even if it results in a no-op. However, the code above does not handle correctly arrays with item-size != 1. When dealing with a memory view rhs I *think* I need to call mp_seq_replace_slice_no_grow() as:
```
src_len = src_slice->len;
src_items = (uint8_t*)src_slice->items + (src_slice->free * item_sz);
...
mp_seq_replace_slice_no_grow((uint8_t*)o->items + (o->free * item_sz), o->len, slice.start, slice.stop, src_items, src_len, item_sz);
```
or replace it with an equivalent memcpy().

dxxb commented:
So you spotted (at least) one issue. Wouldn't it be more straightforward to say what is wrong or provide a complete failing testcase? As much as I enjoy finding things out on my own, I do not always have the time.

This PR does not work with arrays which have an item-size != 1 byte. I extended the tests and fixed the issue but before closing this PR and opening a new one (I imagine this would be the preferred way to do it) I'd like to hear if you spotted anything else besides the mp_seq_replace_slice_no_grow() related comment.

pfalcon commented:
> So you spotted (at least) one issue. Wouldn't it be more straightforward to say what is wrong or provide a complete failing testcase? As much as I enjoy finding things out on my own, I do not always have the time.

With my maintainer's hat on, I likely have even less time. I'm however trying to provide prompt feedback for submissions, so if I see something "strange" (mind the quotes), I'm telling that right away, rather than in a month. If you want specific "wrongs", then you should not use memcpy(), and should use mp_seq_replace_slice_no_grow() - it's there exactly to prevent mistakes like that. Of course, I may be missing something, so I invited you to explain why you did it otherwise.

> but before closing this PR and opening a new one (I imagine this would be the preferred way to do it) 

Nope, fixing this PR is preferred way, that's how it works - people submit PR, get review comments, act on them, rinse an repeat until it's ok to be merged ;-). Thanks.

dbrignoli commented:
> With my maintainer's hat on, I likely have even less time. I'm however trying to provide prompt feedback for submissions, so if I see something "strange" (mind the quotes), I'm telling that right away, rather than in a month. If you want specific "wrongs", then you should not use memcpy(), and should use mp_seq_replace_slice_no_grow() - it's there exactly to prevent mistakes like that. Of course, I may be missing something, so I invited you to explain why you did it otherwise.
> 
ACK. Do we want an assert() or an extra check to make sure the memmove() will be a no-op in case of an lhs memoryview?
> Nope, fixing this PR is preferred way, that's how it works - people submit PR, get review comments, act on them, rinse an repeat until it's ok to be merged ;-). Thanks.
> 
I thought GitHub would lose comments associated with the PR if I reworked it and pushed a new version. I’ll proceed as you say. Thanks!

pfalcon commented:
> Do we want an assert() or an extra check to make sure the memmove() will be a no-op in case of an lhs memoryview?

Sorry, I cannot answer that question right away. The best suggestion I have is the same as said above: "just use mp_seq_replace_slice_no_grow() like existing code does". If you can't use it, then I still don't know why, even though I invited you to elaborate on that.

dxxb commented:
> Sorry, I cannot answer that question right away. The best suggestion I have is the same as said above: "just use mp_seq_replace_slice_no_grow() like existing code does". If you can't use it, then I still don't know why, even though I invited you to elaborate on that.
> 

Hi, I updated the PR a few days ago. I am now using mp_seq_replace_slice_no_grow() and augmented the tests to cover the case you originally mentioned. Thanks

pfalcon commented:
> Hi, I updated the PR a few days ago. I am now using mp_seq_replace_slice_no_grow() and augmented the tests to cover the case you originally mentioned. Thanks

Well, that's something what github doesn't do - automatically notifying of new version of PRs, leaving that to humans ;-).

Ok, what about following testcase:

~~~~
m2[0:4] = m1[1:3]
~~~~

? (in terms of your additions to the test file).

dxxb commented:
I had missed that testcase. PR updated with new testcase and required fix to make it pass. Thanks

dpgeorge commented:
Thanks @dxxb, this is something I've wanted for a while now :)

For the test, you don't need to try/except for cases that are expected to pass.  If a case is not supported then the exception will stop the py script and the run-test script will mark it as failed.  try/except should only be used in tests when the desired outcome is to actually raise an exception (as in the ValueError test that you wrote).

To get Travis to pass you need to add `#define MICROPY_PY_ARRAY_SLICE_ASSIGN (1)` to qemu-arm/mpconfigport.h.

dpgeorge commented on py/objarray.c:
```diff
@@ -397,6 +404,9 @@ STATIC mp_obj_t array_subscr(mp_obj_t self_in, mp_obj_t index_in, mp_obj_t value
                 mp_int_t len_adj = src_len - (slice.stop - slice.start);
                 if (len_adj > 0) {
+                    if (o->base.type == &mp_type_memoryview) {
+                            goto compat_error;
```
Indentation error.

dxxb replied:
Ack, thanks. I'll fix and push an update.

dxxb commented:
Hi!

> For the test, you don't need to try/except for cases that are expected to pass. If a case is not supported then the exception will stop the py script and the run-test script will mark it as failed. try/except should only be used in tests when the desired outcome is to actually raise an exception (as in the ValueError test that you wrote).

Just to make sure we are on the same page: your comment quoted above describes my understanding of how tests are supposed to be written and I believe I added try/except only when I expect an exception to be raised. Did you see something that does not look right in the test cases I added?

> To get Travis to pass you need to add #define MICROPY_PY_ARRAY_SLICE_ASSIGN (1) to qemu-arm/mpconfigport.h.

Thanks, that was going to be my next question :)

dpgeorge commented on tests/basics/memoryview1.py:
```diff
@@ -36,3 +36,67 @@
+except Exception as e:
+    print(e)
+else:
+    print(b2)
```
Eg here: the `m2[1:3] = m1[0:2]` is expected to succeed, and so should not be wrapped in a try block.  This test should simply be 2 lines: the copy and the print(b2).

dxxb replied:
Oh ouch, my bad. I didn't even even remember writing the tests like that. I have to stop doing too many things at the same time :P

dpgeorge commented on tests/basics/memoryview1.py:
```diff
@@ -36,3 +36,67 @@
+except ValueError:
+    print("ValueError")
+except Exception as e:
+    print(e)
```
For here, if you expect the assignment to fail with ValueError then you just need the `except ValueError` block, not also the general `except Exception as e` block.

dxxb replied:
ditto. Will fix.

dpgeorge commented:
Rebased and merged in 32aba40830320867bdddcf4e0a6658b9df49ac9a and cceff157ddaf42e6c6390209ea14b7f4997b32c0.

Thanks @dxxb!

Note that there is room for optimising code size, but that can be left for another day.

dxxb commented:
Hi again. You just beat me to it ;) see https://github.com/dxxb/micropython/tree/dxb/slice_assign_diet
Not surprisingly your version is better. Do you think we could fold the two instances of mp_seq_replace_slice_grow_inplace() into one like I did in my patch?

dpgeorge commented:
Yes, I also considered the second bit of optimisation, but found that it only saved 4 bytes on Thumb2 (I did it a bit differently though, didn't combine the mp_seq_xx code).  If you want to try doing it your way, go ahead!  Please check that it compiles without memoryview enabled (use unix port, disable memoryview definition in unix/mpconfigport.h).

dxxb commented:
Done, please take a look at: https://github.com/micropython/micropython/pull/1386
